### ISSUE TITLE
traefik does not passTLSClientCert on tls session resumption

### ISSUE THREAD

**Issue by kayuma** [traefik/traefik#7644]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Configured middleWare "passTLSClientCert" to be able to retrieve the Client certificate informations on the backend via http header as documented. In my case I want to evaluate these headers on the forwardAuth service.
Works fine as long as no tls session resumption is happening (first request). 

### What did you expect to see?

http headers with the certificate informations on every request:
- X-Forwarded-Tls-Client-Cert
- X-Forwarded-Tls-Client-Cert-Info

### What did you see instead?

Headers are not set when a tls session is reused. I even get log entries in this case:

```
time="2020-12-09T14:49:53Z" level=error msg="Error while Hello: bufio: buffer full"
time="2020-12-09T14:49:53Z" level=warning msg="Tried to extract a certificate on a request without mutual TLS" middlewareName=esb-test-passtlscert@kubernetescrd middlewareType=PassClientTLSCert
```

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
2.3.4
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```yml
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: passtlscert
spec:
  passTLSClientCert:
    pem: false
    info:
      subject:
        commonName: true
        serialNumber: true
      issuer:
        commonName: true
        serialNumber: true
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
time="2020-12-09T14:49:53Z" level=error msg="Error while Hello: bufio: buffer full"
time="2020-12-09T14:49:53Z" level=warning msg="Tried to extract a certificate on a request without mutual TLS" middlewareName=esb-test-passtlscert@kubernetescrd middlewareType=PassClientTLSCert
```

**Comment by jspdown** [comment]
Hello @kayuma :wave: and thanks for your interest in Traefik.

I tried to reproduce the issue but without success. 

> Error while Hello: bufio: buffer full

It tells us that the expected length of the client hello message is smaller than specified in the record.  

In order to diagnose this further we would need more info. Could you give us your complete dynamic config, static config, Traefik logs in debug mode (from the beginning), TLS version?

**Comment by kayuma** [comment]
Hi @jspdown,

Unfortunately I do no longer have the traefik installation and therefore I can't provide all of the requested configs/logs. Sorry about that. But what I can say is that the setup was in no way special. Configure TLSOption like
```
apiVersion: traefik.containo.us/v1alpha1
kind: TLSOption
metadata:
  name: tls-options
spec:
  minVersion: VersionTLS12
  cipherSuites:
    - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
    - TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
    - TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
  clientAuth:
    secretNames:
      - quovadis-europe-ev-ssl-ca-g1
    clientAuthType: RequireAndVerifyClientCert
  sniStrict: true
```
Apply these TLSOption to an IngressRoute:
```
...
  tls:
    options:
      name: tls-options-es.bfh.ch
      namespace: esb-test
...
```

and apply the following middlewares to the IngressRoute:
```
- name: passtlscert
- name: forward-auth-service-v1
```

I expected to see the "X-Forwarded-Tls-Client-Cert-*" headers in every request on the auth-service backend but when a TLS session resumption occurs the headers are missing.

The difficulties to reproduce it will be to trigger a tls session resumption. With the SoapUI tool I could easily trigger a TLS session resumption. 
As a further help to diagnose the issue you can pass the java system property "-Djavax.net.debug=all" to soapui. Java logs then the complete TLS traefik incl. handshake and resumption informations to stdout.

Apart from SoapUI I could even trigger a session resumption with firefox. Just do a full reload the page (pressing the shift key while clicking on reload). When I remember correctly firefox did not resume the session on every reload but most of the time. Whats interesting here is that firefox did a session resumption while holding the "shift" key but not when doing a normal reload. I expected the opposite, but this is a different story.

Hope that helps further,

Marc

**Comment by bngabor** [comment]
I ran traefik tcp/tls router performance test with 70 clients and confirmed the log message: "Error while Hello: bufio: buffer full".
Added manually certificate which issued by LetsEncrypt.
I think related package: https://golang.org/pkg/bufio/#pkg-variables

**Comment by bngabor** [comment]
Increased resources for the client and error goes to away.

**Comment by leitmedium** [comment]
@bngabor Can you share how or at least where you did that?

**Comment by rowi1de** [comment]
@jspdown this issue happened for us as well (among others). Would it make sense to make the max cert size configurable in Traefik or is that more of a workaround for a bug with session resume?

**Comment by bngabor** [comment]
@leitmedium increased cpu and mem for docker engine on client side.

**Comment by ldez** [comment]
closed by #8194
